frontend/src/app/crm/leads/leads.component.ts

```diff
@@ -1,64 +1,173 @@
+  selection = new SelectionModel<Lead>(true, []);
+  /**
+   * Column IDs that we plan to show on the table are stored here
+   */
   readonly displayedColumns: string[] = [
+    'select',
     'lead_id',
     'name',
     'phone_number',
     'campaign_id',
     'date',
+    'details'
   ];
 
   @ViewChild(MatPaginator, {static: true}) paginator: MatPaginator;
   @ViewChild(MatSort, {static: true}) sort: MatSort;
 
-  constructor(private readonly leadService: LeadService) {
+  constructor(private readonly leadService: LeadService, public dialog: MatDialog) {
     this.dataSource = new MatTableDataSource();
-
     this.loadAllLeads();
   }
 
   ngAfterViewInit(): void {
     this.dataSource.paginator = this.paginator;
+
+    /**
+     * This will let the dataSource sort feature to access nested properties in the JSON such as column_data.
+     */
+    this.dataSource.sortingDataAccessor = (lead, property) => {
+      switch(property) {
+        case 'name': return lead.column_data.FULL_NAME;
+        case 'phone_number': return lead.column_data.PHONE_NUMBER;
+        default: return lead[property];
+      }
+    };
     this.dataSource.sort = this.sort;
   }
 
+ /**
+  * This will access the leads from the leadService and handle the filterPredicate and the isLoading boolean value.
+  */
   loadAllLeads(): void {
     this.isLoading$.next(true);
     this.leadService.getAllLeads().pipe(first()).subscribe((leads) => {
       this.dataSource.data = leads;
+
+      /**
+       * @param data The whole data we have in the JSON.
+       * @param filter The value that the user searches transformedFilter.
+       */
+      this.dataSource.filterPredicate = (data, filter: string)  => {
+        const accumulator = (currentTerm, key) => {
+          //Using the recusrsive method here becuase using an if/else condition will increase space complexity n times.
+          return this.nestedPropertyFilterCheck(currentTerm, data, key);
+        };
+        const dataStr = Object.keys(data).reduce(accumulator, '').toLowerCase();
+
+        /**
+         * Transform the filter by converting it to lowercase and removing whitespace.
+         */
+        const transformedFilter = filter.trim().toLowerCase();
+        return dataStr.indexOf(transformedFilter) !== -1;
+      };
       this.isLoading$.next(false);
-    });
+      });
+  }
+
+  /**
+   * A Recursive function to check for each data in the nested JSON
+   * @param leadString A lead that exists in the JSON
+   * @param data The whole data we have in the JSON
+   * @param key The property of the lead at is nested if it exists
+   * @return leadString
+   */
+  nestedPropertyFilterCheck(leadString, data, key) {
+    if (typeof data[key] === 'object') {
+      for (const k in data[key]) {
+        if (data[key][k] !== null) {
+            leadString = this.nestedPropertyFilterCheck(leadString, data[key], k);
+          }
+        }
+    } else {
+        leadString += data[key];
```

Comment:
Maybe here just check if filter is a substring in data[key]

Reply:
What would that improve?

Reply:
instead of concatenating, you can just check each leadString value to see if it "indexOf(transformedFilter) !== -1". If at any point the value is true, you can just return true, otherwise keep recursing.